### PR TITLE
feat: Custom fields in pre-chat form

### LINKED ISSUE #2744: feat: Custom Fields in pre-chat form
### Problem Statement

Right now, our pre-chat form only supports name, email and message. Adding more combinations of fields helps the customers in different business needs. So the customers can completely customize whatever details are required before starting a new conversation.

### Proposed solution

Add support for various combos, 

- [ ] Name and phone number
- [ ] Name, email, phone number
- [ ] Name, phone number, message

### Design Mocks

Screenshot 2022-03-02 at 12 58 23 PM
Screenshot 2022-03-02 at 12 58 51 PM
Screenshot 2022-03-02 at 12 59 10 PM

_Originally posted by @tabrza in https://github.com/chatwoot/chatwoot/issues/1131#issuecomment-891577370_

### PR DESCRIPTION
**Pull request by muhsin-k**
Fixes #2744 

- [x] Added [vueformulate](https://vueformulate.com/) for dynamic pre-chat form
- [x] Added support for rendering types Text, Number, Link, Date, List, Checkbox
- [x] Added the option to collect custom attributes via pre-chat form

### CODE REVIEW

**Comment by iamsivin** [comment]
Tested locally and found some issues.
1. Input labels are not visible and the input colour is not correct in dark mode.
2. Pre chat message is not shown in the widget.
3. Email address toggle button is working if we click on other fields.
4. There is now a warning showing for custom attribute type `number`.

Can see it in the bellow screencast.

https://user-images.githubusercontent.com/64252451/161704286-abdbed1d-72fc-4c31-8eb5-b7f0eb735a75.mov

**Comment by muhsin-k** [comment]
> Tested locally and found some issues.
> 
> * [x]  Input labels are not visible and the input colour is not correct in dark mode.
> * [x]  Pre chat message is not shown in the widget.
> * [x]  There is no warning showing for custom attribute type `number`.
> * [x]  Email address toggle button is working if we click on other fields.
> 
> Can see it in the bellow screencast. Had a call with @muhsin-k
> 
>  Screen.Recording.2022-04-05.at.1.09.39.PM.mov

All these issues are fixed now.

**Comment by pranavrajs** [comment]
- Create a Table Row Component for rendering the inputs, PreChatSettings Component is bloated right now.
- There is no need to support old object attributes in the UI. If migration is successful, then the rest of the code would never be executed.

**Review by sojan-official — APPROVED** [review]

**Review by iamsivin — APPROVED** [review]
Tested locally LGTM⚡️

File: app/javascript/widget/i18n/locale/en.json [hunk 1]
```diff
@@ -22,7 +22,6 @@
     "IN_A_DAY": "Typically replies in a day"
   },
   "START_CONVERSATION": "Start Conversation",
-  "END_CONVERSATION": "End Conversation",
```
Comment (iamsivin) [inline app/javascript/widget/i18n/locale/en.json:25]: @muhsin-k Can you revert back this line? 

**Screenshot**

image

**Review by iamsivin — APPROVED** [review]

**Comment by muhsin-k** [comment]
@iamsivin Could you please do the QA for all the issues reported in https://github.com/chatwoot/chatwoot/issues/4469

**Comment by iamsivin** [comment]
Tested locally

1. Sync pre-chat fields after custom attribute destroy -> Tested looks good to me.

2. When setUser is used, pre-chat forms are not visible -> Tested found an issue, after using setUser the sent message API call is falling you can see in this loom video.
https://www.loom.com/share/b5da6369190e4e00b3f7eb6a16861bb4

3. Dark mode colors - Checkbox bg and Date Icon -> Tested LGTM 
https://user-images.githubusercontent.com/64252451/163797305-d325b97f-7b84-4563-8445-dcc78953cf4b.mov

4. After setUser, user information is asked again in the pre-chat form if the campaign is clicked -> Not tested

**Review by nithindavid — APPROVED** [review]

**Review by iamsivin — APPROVED** [review]

**Pull request merged**